From 36b294f7b9eff808fbbc4f4dde377b375f803601 Mon Sep 17 00:00:00 2001
From: Abhishek Ojha <abhishek.ojha@savoirfairelinux.com>
Date: Wed, 24 Dec 2025 14:35:52 -0500
Subject: [PATCH] pci: add msi-x interrupt support

---
 hw/pci/minimal_pcie_nic.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/hw/pci/minimal_pcie_nic.c b/hw/pci/minimal_pcie_nic.c
index 773577661..a9b14c24f 100644
--- a/hw/pci/minimal_pcie_nic.c
+++ b/hw/pci/minimal_pcie_nic.c
@@ -22,13 +22,16 @@
 #include "hw/irq.h"
 
 #define TYPE_MINIMAL_PCIE_NIC "minimal-pcie-nic"
+#define MSIX_NUM_VECTORS 1
+#define MSIX_BAR_SIZE   0x1000
 
 OBJECT_DECLARE_SIMPLE_TYPE(MinimalPCIeNICState, MINIMAL_PCIE_NIC)
 
 /* Device state structure */
 typedef struct MinimalPCIeNICState {
     PCIDevice parent_obj;      /* Must be first */
-    MemoryRegion mmio;         /* BAR0 MMIO region (trapped by callbacks) */
+    MemoryRegion mmio;         /* BAR0 Device Registers */
+    MemoryRegion msix_bar;      /* BAR1: MSI-X table + PBA */
     uint32_t regs[16];         /* Simulated device registers (64 bytes) */
 } MinimalPCIeNICState;
 
@@ -147,10 +150,23 @@ static void minimal_pcie_nic_realize(PCIDevice *pdev, Error **errp)
      * Guest OS will map this BAR, reads/writes hit callbacks
      */
     pci_register_bar(pdev, 0, PCI_BASE_ADDRESS_SPACE_MEMORY, &s->mmio);
+
+        /* BAR1: MSI-X table */
+    memory_region_init(&s->msix_bar, OBJECT(s),"minimal-msix-bar", MSIX_BAR_SIZE);
+
+    pci_register_bar(pdev, 1, PCI_BASE_ADDRESS_SPACE_MEMORY, &s->msix_bar);
+      /* MSI-X capability */
+    msix_init(pdev, MSIX_NUM_VECTORS, &s->msix_bar,
+              1,                  /* BAR index */
+              0x0,                /* table offset */
+              &s->msix_bar,
+              1,
+              0x800);             /* PBA offset */
 }
 
 static void minimal_pcie_nic_uninit(PCIDevice *pdev)
 {
+    msix_uninit(pdev);
     printf("pcie nic un-init\n");
 }
 
