From bfdbe203d930886464de5a1b6c63f42d8a60232c Mon Sep 17 00:00:00 2001
From: Abhishek Ojha <abhishek.ojha@savoirfairelinux.com>
Date: Sun, 11 Jan 2026 17:14:16 -0500
Subject: [PATCH] pcie: add rx path packet processing to test the packet flow

---
 hw/pci/minimal_pcie_nic.c | 132 +++++++++++++++++++++++++++++++++++---
 1 file changed, 124 insertions(+), 8 deletions(-)

diff --git a/hw/pci/minimal_pcie_nic.c b/hw/pci/minimal_pcie_nic.c
index a2c5140fe..16e78eba2 100644
--- a/hw/pci/minimal_pcie_nic.c
+++ b/hw/pci/minimal_pcie_nic.c
@@ -20,13 +20,15 @@
 #include "qapi/error.h"
 #include "exec/memory.h" /* MemoryRegion */
 #include "hw/irq.h"
+#include "net/net.h"
 
 #define TYPE_MINIMAL_PCIE_NIC "minimal-pcie-nic"    // qemu device name
 #define MSI_NUM_VECTORS         4                   // msi max vectors
 #define MSIX_NUM_VECTORS        4                   // msi-x max vectors
 #define MSIX_BAR_SIZE           0x1000              // 4KB MSIX Bar size
 #define MSIX_ENABLE                                 // Select MSI or MSI-X
-#define MSIX_IDX                1                   // Use BAR 1 for MSI-X
+#define BAR1_MSIX_IDX           1                   // Use BAR 1 for MSI-X
+#define BAR0_IDX                0                   // Use BAR 0 for MMIO
 
 OBJECT_DECLARE_SIMPLE_TYPE(MinimalPCIeNICState, MINIMAL_PCIE_NIC)
 
@@ -36,8 +38,35 @@ typedef struct MinimalPCIeNICState {
     MemoryRegion mmio;         /* BAR0 Device Registers */
     MemoryRegion msix_bar;      /* BAR1: MSI-X table + PBA */
     uint32_t regs[16];         /* Simulated device registers (64 bytes) */
+
+    NICState *nic;
+    NICConf conf;
+    NetClientState *nc;
+
+    uint64_t rx_ring_base;
+    uint32_t rx_ring_size;
+    uint32_t rx_head;
+    uint32_t rx_tail;
 } MinimalPCIeNICState;
 
+#define REG_RX_RING_BASE   0x10
+#define REG_RX_RING_SIZE   0x18
+#define REG_RX_TAIL        0x1C
+#define REG_RX_HEAD        0x20
+
+struct rx_desc {
+    uint64_t addr;
+    uint16_t len;
+    uint16_t flags;
+};
+
+#define RX_DONE 1
+
+/* Callback; packet received from host (eg. tap or user networking) */
+static ssize_t minimal_receive_packet(NetClientState *nc,
+                                      const uint8_t *buf,
+                                      size_t size);
+
 /* Generate MSI/MSI-X interrupt */
 static void minimal_raise_irq(MinimalPCIeNICState *s, uint32_t vector)
 {
@@ -70,10 +99,10 @@ minimal_init_msix(MinimalPCIeNICState *s)
 
     res = msix_init(PCI_DEVICE(s), MSIX_NUM_VECTORS,
                     &s->msix_bar,
-                    MSIX_IDX,       /* BAR index */
-                    0,              /* table offset */
+                    BAR1_MSIX_IDX,       /* BAR index */
+                    0,                  /* table offset */
                     &s->msix_bar,
-                    MSIX_IDX,       /* PBA BAR index */
+                    BAR1_MSIX_IDX,       /* PBA BAR index */
                     0x800,
                     0x98, NULL);
 
@@ -133,6 +162,21 @@ static void minimal_mmio_write(void *opaque,
 {
     MinimalPCIeNICState *s = opaque;
 
+    if (addr == REG_RX_RING_BASE) {
+        s->rx_ring_base = data;
+        return;
+    }
+
+    if (addr == REG_RX_RING_SIZE) {
+        s->rx_ring_size = data;
+        return;
+    }
+
+    if (addr == REG_RX_TAIL) {
+        s->rx_tail = data;
+        return;
+    }
+
     if (addr + size > sizeof(s->regs)) {
         printf("minimal_pcie_nic: MMIO write out-of-bounds addr=0x%#" PRIx64
                " size=%u\n", (uint64_t)addr, size);
@@ -184,10 +228,56 @@ static const MemoryRegionOps minimal_mmio_ops = {
     .endianness = DEVICE_NATIVE_ENDIAN,
 };
 
+static ssize_t minimal_receive_packet(NetClientState *nc,
+                                      const uint8_t *buf,
+                                      size_t size)
+{
+    MinimalPCIeNICState *s = qemu_get_nic_opaque(nc);
+    struct rx_desc desc;
+    uint64_t desc_addr;
+
+    if (!s->rx_ring_size)
+        return 0;   // driver not ready
+
+    desc_addr = s->rx_ring_base +
+                s->rx_head * sizeof(desc);
+
+    /* Read RX descriptor */
+    pci_dma_read(&s->parent_obj,
+                 desc_addr, &desc, sizeof(desc));
+
+    /* DMA packet into guest memory */
+    pci_dma_write(&s->parent_obj,
+                  desc.addr, buf, size);
+
+    /* Update descriptor */
+    desc.len = size;
+    desc.flags = RX_DONE;
+
+    pci_dma_write(&s->parent_obj,
+                  desc_addr, &desc, sizeof(desc));
+
+    /* Advance ring */
+    s->rx_head = (s->rx_head + 1) % s->rx_ring_size;
+
+    /* Fire MSI-X vector 0 */
+    minimal_raise_irq(s, 0);
+    printf("minimal_pcie_nic: MSI-X interrupt triggered\n");
+
+    return size;
+}
+
+static NetClientInfo net_ops = {
+    .type = NET_CLIENT_DRIVER_NIC,
+    .size = sizeof(NICState),
+    .receive = minimal_receive_packet,
+};
+
 /* Realize function: called when device is instantiated */
 static void minimal_pcie_nic_realize(PCIDevice *pdev, Error **errp)
 {
     MinimalPCIeNICState *s = MINIMAL_PCIE_NIC(pdev);
+    uint8_t *macaddr;
 
     printf("minimal_pcie_nic: realize called (host log)\n");
 
@@ -219,13 +309,12 @@ static void minimal_pcie_nic_realize(PCIDevice *pdev, Error **errp)
     /* Register BAR0 with PCI core
      * Guest OS will map this BAR, reads/writes hit callbacks
      */
-    pci_register_bar(pdev, 0, PCI_BASE_ADDRESS_SPACE_MEMORY, &s->mmio);
+    pci_register_bar(pdev, BAR0_IDX , PCI_BASE_ADDRESS_SPACE_MEMORY, &s->mmio);
 
     /* BAR1: MSI-X table */
-    memory_region_init(&s->msix_bar, OBJECT(s), "minimal-msix-bar", 
-                        MSIX_BAR_SIZE);
+    memory_region_init(&s->msix_bar, OBJECT(s), "minimal-msix-bar", MSIX_BAR_SIZE);
 
-    pci_register_bar(pdev, 1, PCI_BASE_ADDRESS_SPACE_MEMORY, &s->msix_bar);
+    pci_register_bar(pdev, BAR1_MSIX_IDX, PCI_BASE_ADDRESS_SPACE_MEMORY, &s->msix_bar);
 
 #ifdef MSIX_ENABLE
     /* MSI-X capability */
@@ -242,11 +331,29 @@ static void minimal_pcie_nic_realize(PCIDevice *pdev, Error **errp)
     }
 #endif
 
+    qemu_macaddr_default_if_unset(&s->conf.macaddr);
+    macaddr = s->conf.macaddr.a;
+
+    /* Create NIC backend */
+    s->nic = qemu_new_nic(&net_ops,
+                      &s->conf,                    // default config
+                      object_get_typename(OBJECT(s)),      // NIC name
+                      DEVICE(pdev)->id,        // unique NIC ID
+                      &PCI_DEVICE(s)->qdev.mem_reentrancy_guard,
+                      s);
+
+    qemu_format_nic_info_str(qemu_get_queue(s->nic), macaddr);
 
 }
 
 static void minimal_pcie_nic_uninit(PCIDevice *pdev)
 {
+    MinimalPCIeNICState *s = MINIMAL_PCIE_NIC(pdev);
+
+    /* Clean up NIC */
+    qemu_del_nic(s->nic);
+
+    /* Clean up MSI/MSI-X */
 #ifdef MSIX_ENABLE
     msix_uninit(pdev);
 #else
@@ -255,6 +362,12 @@ static void minimal_pcie_nic_uninit(PCIDevice *pdev)
     printf("pcie nic un-init\n");
 }
 
+/* Device properties */
+static Property minimal_pcie_nic_properties[] = {
+    DEFINE_NIC_PROPERTIES(MinimalPCIeNICState, conf),
+    DEFINE_PROP_END_OF_LIST(),
+};
+
 /* Class initialization */
 static void minimal_pcie_nic_class_init(ObjectClass *klass, void *data)
 {
@@ -271,6 +384,9 @@ static void minimal_pcie_nic_class_init(ObjectClass *klass, void *data)
     k->revision = 0x1; //set initial revision
     set_bit(DEVICE_CATEGORY_NETWORK, dc->categories);
     dc->desc = "Minimal PCIe NIC Card";
+    
+    /* Register device properties (netdev, mac, etc.) */
+    device_class_set_props(dc, minimal_pcie_nic_properties);
 }
 
 /* Type registration */
